### PR TITLE
Add Inventory Operating System for VMware ESX Server through vCenter

### PR DESCRIPTION
**Pull request by NimVek**
Setup:
Use "agent_vsphere" to check vCenter Server, direct connection to ESX Server is not possible.

Current behavior:
Because there is no piggiback section "esx_systeminfo" generated for each esx server registered in vcenter server,
they are no information about the os and no label for "cmk/vsphere_object:server" is generated.

Behavior after patch:
In section "esx_vsphere_hostsystem" additional information about os of esx server is gathered.
If section "esx_vsphere_hostsystem" hostlabel "cmk/vsphere_object:server" is generated.

Why:
If the cmk site has no direct connection to the esx server, it cant use the "agent_vsphere" to gather os information.

p.s.:
The use of the host label "cmk/vsphere_object" maybe could have some bugs, because a vcenter or esx server could be a vm at the same time but a host cant have the labels "cmk/vsphere_object:vcenter" and "cmk/vsphere_object:vm" at the same time.
If you like i can add another PR, so labels like "cmk/vsphere_vcenter:yes", "cmk/vsphere_esx:yes" or "cmk/vsphere_vm:yes"
are generated.

### CODE REVIEW

**Comment by NimVek** [comment]
Because you revoked parts of PR https://github.com/tribe29/checkmk/pull/339 an additional substantiation in german.

Ich hatte in meinem früheren Pull Request nicht ganz herausgearbeitet, das es mir darum ging die Informationen über das Betriebssystem des ESX-Servers aus den Informationen des vCenter-Servers zu erlangen.
In unserem Aufbau ist es nicht möglich die ESX-Server direkt abzufragen, somit ist der Umweg über das vCenter unerlässlich.
Die angesprochene Lösung über "esx_systeminfo" ist nicht möglich, da der Agent "agent_vsphere" nur einmal diese Sektion erzeugt und auch nur für den Host mit dem er direkt verbunden ist. In unserem Fall also nur das vCenter. Die Sektionen für die ESX-Server werden nicht per Piggyback erzeugt, sondern könnten auch nur mittels direkter Kommunikation erzeugt werden.
Es wäre sicher möglich den "agent_vsphere" so zu ändern, dass er per piggyback die Sektion "esx_systeminfo" für jeden ESX-Server erzeugt, aber dies würde den Programmablauf komplizierter gestallten und wäre nicht so elegant, deshalb habe ich die Lösung über die Sektion  "esx_vsphere_hostsystem" gewählt.

**Comment by NimVek** [comment]
In my environment i get an failure in PR-CI Workflow:
"testing (mypy tests, test-mypy-raw)" fails on cmk/special_agents/agent_kube.py, but i never touched this file.
I hope its ok?
If not, i could resolve the issue, but its would mix different issues in one Pull request.

File: cmk/base/plugins/agent_based/esx_vsphere_hostsystem_section.py [hunk 1]
```diff
@@ -34,7 +34,21 @@ def parse_esx_vsphere_hostsystem(string_table: StringTable) -> Section:
...
+    """Host label function
+
+    Labels:
+
+        cmk/vsphere_object:
+            This label is set to "vcenter" if the corresponding host is a
+            VMWare vCenter, and to "server" if the host is an ESXi hostsystem.
+
+    """
+    yield HostLabel("cmk/vsphere_object", "server")
```
Comment (kain88-de) [inline cmk/base/plugins/agent_based/esx_vsphere_hostsystem_section.py:47]: Please change the docstring to reflect the behavior of the function. Ideally use a sentence why it makes sense that this only ever returns this label.

**Comment by NimVek** [comment]
I added additional info to the host_label_function.
The description of the Label is the same as in esx_systeminfo, because it describes the possible values, and because of "test_all_sections_have_host_labels_documented" it must be the same in every occurence.

**Comment by kain88-de** [comment]
Thank you for the description. I had a quick chat with a colleague who understands vSphere better than me. This change breaks existing configs, as we potentially gather host information from multiple sources (vCenter and ESXi servers). Your suggested alternative of adding `esx_systeminfo` to the piggyback sections of ESXi servers can break configs as well. I'm checking with our product management if we want to merge this, likely with some changes. 

Since this is a breaking change, if we merge this change it will only be done in master, and released with 2.2. This will not be ported to 2.1. or earlier.

In the mean time. You can create an MKP of this version of the vsphere agent and checks, and publish on our [exchange](https://exchange.checkmk.com/) if you like.

**Comment by kain88-de** [comment]
This will be merged shortly. Thanks for your contribution.

**Pull request closed without merging**